### PR TITLE
Add logging

### LINKED ISSUE #40: Logging
Logging to serial. Should support multiple levels (debug, warning, info and error). An example log message is `[DEBUG] This is the message`.

The logging interface should provide a function for each level which all call a generic function, and the generic function itself.

### PR DESCRIPTION
**Pull request by SamTebbs33**
This patch adds serial logging on top of the serial interface from #41. Closes #40. Depends on #41

### CODE REVIEW

File: src/kernel/log.zig [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    WARNING,
+    ERROR
+};
+
+pub fn init() void {
+    serial.init(serial.DEFAULT_BAUDRATE, serial.Port.COM1) catch unreachable;
+}
+
+fn logCallback(context: void, str: []const u8) anyerror!void {
+    for (str) |char| {
```
Comment (DrDeano) [inline src/kernel/log.zig]: U have a writeStr in ur serial interface

File: src/kernel/log.zig [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+pub const Level = enum {
+    INFO,
+    DEBUG,
+    WARNING,
+    ERROR
+};
+
+pub fn init() void {
+    serial.init(serial.DEFAULT_BAUDRATE, serial.Port.COM1) catch unreachable;
```
Comment (DrDeano) [inline src/kernel/log.zig]: Its probs best to init the serial separately as other stuff may use the serial without the logging

File: src/kernel/log.zig (hunk 1 above)
Comment (SamTebbs33) [reply src/kernel/log.zig]: Indeed. This is from before that was added.

File: src/kernel/log.zig (hunk 2 above)
Comment (SamTebbs33) [reply src/kernel/log.zig]: Very true

File: src/kernel/arch/x86/arch.zig [hunk 3]
```diff
@@ -43,7 +44,9 @@ pub const InterruptContext = struct {
 pub fn init() void {
     disableInterrupts();
 
+    log.log_info("Init gdt\n");
```
Comment (SamTebbs33) [inline src/kernel/arch/x86/arch.zig]: I'll move these log messages to the init function they apply to

File: src/kernel/log.zig [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    for (str) |char| {
+        serial.write(char, serial.Port.COM1);
+    }
+}
+
+pub fn log(comptime level: Level, comptime format: []const u8, args: ...) void {
+    fmt.format({}, anyerror, logCallback, "[" ++ @tagName(level) ++ "] " ++ format, args) catch unreachable;
+}
+
+pub fn log_info(comptime format: []const u8, args: ...) void {
```
Comment (SamTebbs33) [inline src/kernel/log.zig]: I'll make these camel case to match the serial functions

**Pull request merged**